lib/smack/IntegerOverflowChecker.cpp

```diff
@@ -0,0 +1,159 @@
+  } else {
+    return APInt::getMinValue(bits).toString(10, false);
+  }
+}
+
+Value* IntegerOverflowChecker::extValue(Value* v, int bits, bool isSigned, Instruction* i, bool check) {
+  if (check) {
+    if (isSigned) {
+      return CastInst::CreateSExtOrBitCast(v, IntegerType::get(i->getFunction()->getContext(), bits*2), "", i);
+    } else {
+      return CastInst::CreateZExtOrBitCast(v, IntegerType::get(i->getFunction()->getContext(), bits*2), "", i);
+    }
+  } else
+    return v;
+}
+
+BinaryOperator* IntegerOverflowChecker::mkFlag(Value* v, int bits, bool isSigned, Instruction* i, bool check) {
+  if (check) {
+    ConstantInt* max = ConstantInt::get(IntegerType::get(i->getFunction()->getContext(), bits*2), getMax(bits, isSigned), 10);
+    ConstantInt* min = ConstantInt::get(IntegerType::get(i->getFunction()->getContext(), bits*2), getMin(bits, isSigned), 10);
+    CmpInst::Predicate maxCmpPred = (isSigned ? CmpInst::ICMP_SGT : CmpInst::ICMP_UGT);
+    CmpInst::Predicate minCmpPred = (isSigned ? CmpInst::ICMP_SLT : CmpInst::ICMP_ULT);
+    ICmpInst* gt = new ICmpInst(i, maxCmpPred, v, max, "");
+    ICmpInst* lt = new ICmpInst(i, minCmpPred, v, min, "");
+    return BinaryOperator::Create(Instruction::Or, gt, lt, "", i);
+  } else {
+    ConstantInt* a = ConstantInt::getFalse(i->getFunction()->getContext());
+    return BinaryOperator::Create(Instruction::And, a, a, "", i);
+  }
+}
+
+Value* IntegerOverflowChecker::mkResult(Value* v, int bits, Instruction* i, bool check) {
+  if (check)
+    return CastInst::CreateTruncOrBitCast(v, IntegerType::get(i->getFunction()->getContext(), bits), "", i);
+  else
+    return v;
+}
+
+void IntegerOverflowChecker::mkCheck(Function* co, BinaryOperator* flag, Instruction* i) {
+  ArrayRef<Value*> args(CastInst::CreateIntegerCast(flag, co->arg_begin()->getType(), false, "", i));
+  CallInst::Create(co, args, "", i);
+}
+
+void IntegerOverflowChecker::mkBlockingAssume(Function* va, BinaryOperator* flag, Instruction* i) {
+  ArrayRef<Value*> args(CastInst::CreateIntegerCast(BinaryOperator::CreateNot(flag, "", i),
+        va->arg_begin()->getType(), false, "", i));
+  CallInst::Create(va, args, "", i);
+}
+
+bool IntegerOverflowChecker::runOnModule(Module& m) {
+  Function* co = m.getFunction("__SMACK_check_overflow");
+  assert(co != NULL && "Function __SMACK_check_overflow should be present.");
+  Function* va = m.getFunction("__VERIFIER_assume");
+  assert(va != NULL && "Function __VERIFIER_assume should be present.");
+  std::vector<Instruction*> instToRemove;
+  for (auto& F : m) {
+    if (!Naming::isSmackName(F.getName())) {
+      for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
+        if (auto ei = dyn_cast<ExtractValueInst>(&*I)) {
+          if (auto ci = dyn_cast<CallInst>(ei->getAggregateOperand())) {
+            Function* f = ci->getCalledFunction();
+            SmallVectorImpl<StringRef> *ar = new SmallVector<StringRef, 3>;
```

Comment:
What does `ar` stand for? Expand this name.

Reply:
Done.